src/sumcheck/sumcheck_small_value.rs

```diff
@@ -0,0 +1,841 @@
+    // L_0 (x) = 1 - x
+    // L_1 (x) = x
+    // L_inf (x) = (x - 1)x
+    let lagrange_evals_r_1 = [-r_1 + F::ONE, r_1];
+
+    // ------------------  Round 2  ------------------ 
+
+    // 1. For u in {0, 1} compute t_2(u).
+    // First we take the accumulators A_2(v, u).
+    // We computed 4 accumulators for round 2, since only need v in {0, 1} and u in {0, 1}.
+    let accumulators_round_2 = accumulators.get_accumulators_for_round(1);
+
+    let mut t_2_evals = [EF::ZERO; 2];
+
+    
+    t_2_evals[0] += lagrange_evals_r_1[0] * accumulators_round_2[0];
+    t_2_evals[0] += lagrange_evals_r_1[1] * accumulators_round_2[2];
+
+    t_2_evals[1] += lagrange_evals_r_1[0] * accumulators_round_2[1];
+    t_2_evals[1] += lagrange_evals_r_1[1] * accumulators_round_2[3];
+
+    // We compute l_2(0) and l_2(inf)
+    let linear_2_evals = compute_linear_function(&w.0[..2], &[r_1]);
+
+    // We compute S_2(0) and S_2(inf).
+    let round_poly_evals = get_evals_from_l_and_t(&linear_2_evals, &t_2_evals);
+
+    // 3. Send S_2(u) to the verifier.
+    prover_state.add_extension_scalars(&round_poly_evals);
+
+    // 4. Receive the challenge r_2 from the verifier.
+    let r_2: EF = prover_state.sample();
+
+    // 5. Compute R_3 = [L_00(r_1, r_2), L_01(r_1, r_2), ..., L_{inf inf}(r_1, r_2)]
+    // L_00 (x1, x2) = (1 - x1) * (1 - x2)
+    // L_01 (x1, x2) = (1 - x1) * x2
+    // ...
+    // L_{inf inf} (x1, x2) = (x1 - 1) * x1 * (x2 - 1) * x2
+
+    let [l_0, l_1] = lagrange_evals_r_1;
+    let one_minus_r_2 = -r_2 + F::ONE;
+
+    let lagrange_evals_r_2 = [
+        l_0 * one_minus_r_2, // L_0 0
+        l_0 * r_2,           // L_0 1
+        l_1 * one_minus_r_2, // L_1 0
+        l_1 * r_2,           // L_1 1
+    ];
+
+    let eval_1 = *sum - round_poly_evals[0];
+    *sum = round_poly_evals[1] * r_2.square()
+        + (eval_1 - round_poly_evals[0] - round_poly_evals[1]) * r_2
+        + round_poly_evals[0];
+
+    // ------------------  Round 3  ------------------ 
+
+    // 1. For u in {0, 1} compute t_3(u).
+
+    // First we take the accumulators A_3(v, u).
+    // We computed 4 accumulators at the third round for v in {0, 1}^2 and u in {0, 1}.
+    let accumulators_round_3 = accumulators.get_accumulators_for_round(2);
+
+    let mut t_3_evals = [EF::ZERO; 2];
+
+    t_3_evals[0] += lagrange_evals_r_2[0] * accumulators_round_3[0]; // (v=00 u=0)
+    t_3_evals[0] += lagrange_evals_r_2[1] * accumulators_round_3[2]; // (10 0)
+    t_3_evals[0] += lagrange_evals_r_2[2] * accumulators_round_3[4]; // (01 0)
+    t_3_evals[0] += lagrange_evals_r_2[3] * accumulators_round_3[6]; // (11 0)
+
+    t_3_evals[1] += lagrange_evals_r_2[0] * accumulators_round_3[1]; // (00 1)
+    t_3_evals[1] += lagrange_evals_r_2[1] * accumulators_round_3[3]; // (01 1)
+    t_3_evals[1] += lagrange_evals_r_2[2] * accumulators_round_3[5]; // (10 1)
+    t_3_evals[1] += lagrange_evals_r_2[3] * accumulators_round_3[7]; // (11 1)
```

Comment:
Here I guess we should be able to compute in one shot as well no?

```suggestion
    let t_3_evals = [
        // t_3(u=0) = Σ_{v} L_v(r_1, r_2) * A(v, u=0)
        lagrange_r2.iter().zip(accumulators_round_3.iter().step_by(2)).map(|(l, a)| *l * *a).sum(),
        // t_3(u=1) = Σ_{v} L_v(r_1, r_2) * A(v, u=1)
        lagrange_r2.iter().zip(accumulators_round_3.iter().skip(1).step_by(2)).map(|(l, a)| *l * *a).sum(),
    ];
```